Use correct boxed_forward_device_index when running `CompiledFxGraph.post_compile`

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148130

This PR threads through the correct boxed_forward_device_index from graph_kwargs to CompiledFXGraph.post_compile. This allows us to correctly update BoxedDeviceIndex from cache hits.

We don't actually need to save `boxed_forward_device_index` in CompiledFXGraph because its value is in the cache key, so it always matches to the ambient one anyway. On forward with cudagraphs enabled, derive `boxed_forward_device_index`'s value from `device_idxs`.

Testing: 

```
python benchmarks/dynamo/cachebench.py --mode training --benchmark torchbench --model BERT_pytorch --device cuda --repeat 1 --dynamic --output="dynamic.json"
```

Now cache hits properly on FXGraphCache. AOTAutogradCache has a guard failure. Will look into that as a followup. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

jamesjwu committed:
Use correct boxed_forward_device_index test



Differential Revision: [D70353454](https://our.internmc.facebook.com/intern/diff/D70353454/)

[ghstack-poisoned]

jamesjwu committed:
Update on "Use correct boxed_forward_device_index test"


Differential Revision: [D70353454](https://our.internmc.facebook.com/intern/diff/D70353454/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu committed:
Update on "Use correct boxed_forward_device_index test"


Differential Revision: [D70353454](https://our.internmc.facebook.com/intern/diff/D70353454/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu commented on torch/_inductor/output_code.py:
```diff
@@ -147,7 +148,6 @@ def cudagraph_post_compile(
     assert compiled_graph.cudagraph_info is not None
     cached_info = compiled_graph.cudagraph_info
     cudagraph_fail_reasons = cached_info.cudagraph_fail_reasons
-    boxed_forward_device_index = compiled_graph.boxed_forward_device_index
```
Here's my logic for removing boxed_forward_device_index from compile_fx_graph. @eellison, please double check my reasoning: 

- `boxed_forward_device_index`'s value should always be set to `next(iter(compiled_graph.device_idxs))` on the forward, and the previous box_forward_device_index on the backward. 

- On a cache miss, `boxed_forward_device_index` should be the value of the forward_device passed to `compile_fx_inner`. Thus, there's no need to save it on the CompiledFxGraph: we should be using the one defined in compile_fx and passed into compile_fx_inner. 

`boxed_forward_device_index` is part of FXGraphCache's key. So on a forward cache hit, we want to set the BoxedDevice passed into compile_fx, and on a backward cache hit, we want to use the BoxedDevice that's already been set by the forward running before (there's also no scenario where the backward could run before the forward, so you're guaranteed on backward cache hit for boxed_forward_device_index to already be set by the forward). 

Thus I see no reason to even store boxed_forward_device_index in the compiled fx graph: we should instead always use the BoxedDevice passed into compile_fx in the first place.

eellison replied:
Yes, it should always be known statically.

eellison replied:
`next(iter(compiled_graph.device_idxs))` - this will no longer be true with cudagraph splitting. but anyway, we'll cross that hurdle when we get there @BoyuanFeng

jamesjwu committed:
Update on "Use correct boxed_forward_device_index test"


Differential Revision: [D70353454](https://our.internmc.facebook.com/intern/diff/D70353454/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu committed:
Update on "Use correct boxed_forward_device_index test"


Differential Revision: [D70353454](https://our.internmc.facebook.com/intern/diff/D70353454/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu committed:
Update on "Use correct boxed_forward_device_index test"


Differential Revision: [D70353454](https://our.internmc.facebook.com/intern/diff/D70353454/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu committed:
Update on "Use correct boxed_forward_device_index test"


This PR threads through the correct boxed_forward_device_index from graph_kwargs to CompiledFXGraph.post_compile. This allows us to correctly update BoxedDeviceIndex from cache hits.

We don't actually need to save `boxed_forward_device_index` in CompiledFXGraph because its value is in the cache key, so it always matches to the ambient one anyway. On forward with cudagraphs enabled, derive `boxed_forward_device_index`'s value from `device_idxs`.


Differential Revision: [D70353454](https://our.internmc.facebook.com/intern/diff/D70353454/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu commented:
Unlinked this from phabricator/internal, will use diff train to land internally

jamesjwu committed:
Update

[ghstack-poisoned]

eellison commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -390,9 +403,17 @@ def load(self, example_inputs, fx_config: _CompileFxKwargs) -> CompiledFxGraph:
             },
             payload_fn=lambda: json.dumps(cache_info),
         )
+        self.example_inputs = example_inputs
```
why do we need to do this ?

jamesjwu replied:
It makes it so we can pass it to the post_compile() below. The goal here is to split the post_compile from the cache loading, so that we don't set the global constants until *after* we're sure the cache successfully hit on both

eellison commented on torch/_inductor/output_code.py:
```diff
@@ -497,6 +493,7 @@ def post_compile(
                     self,
                     cudagraphs,
                     constants.unwrap(self),
+                    graph_kwargs.get("boxed_forward_device_index", None),
```
could we add assertions here - do we know if this is fwd/backward ? if we know its backward, whenever cudagraphs is set this should also be set..

jamesjwu replied:
Yup will do

eellison reviewed (commented):
Can you add a test for this ? Also, would be good to add a cache test equivalent of https://github.com/pytorch/pytorch/blob/83fb974b5d1ec8c31f2da7e9c5e4794ff8d767a4/test/inductor/test_cudagraph_trees.py#L596 which ensures that if we ran the forward in cudagraphs, the backward `set_to_running_backward` gets called. maybe we already did add this, i'm not sure.

jamesjwu commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -379,6 +383,15 @@ def load(self, example_inputs, fx_config: _CompileFxKwargs) -> CompiledFxGraph:
         )
         if result is None:
             log.info("FXGraphCache cache miss for key %s", self.fx_graph_cache_key)
+            torch._logging.trace_structured(
```
cc @masnesral

jamesjwu committed:
Add a test to test_cudagraph_trees, add assertions on "Use correct boxed_forward_device_index when running `CompiledFxGraph.post_compile`"


This PR threads through the correct boxed_forward_device_index from graph_kwargs to CompiledFXGraph.post_compile. This allows us to correctly update BoxedDeviceIndex from cache hits.

We don't actually need to save `boxed_forward_device_index` in CompiledFXGraph because its value is in the cache key, so it always matches to the ambient one anyway. On forward with cudagraphs enabled, derive `boxed_forward_device_index`'s value from `device_idxs`.

Testing: 

```
python benchmarks/dynamo/cachebench.py --mode training --benchmark torchbench --model BERT_pytorch --device cuda --repeat 1 --dynamic --output="dynamic.json"
```

Now cache hits properly on FXGraphCache. AOTAutogradCache has a guard failure. Will look into that as a followup. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jamesjwu commented:
Added a test to show that aotautograd cache hits + we correctly use the proper device index

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu commented:
@pytorchbot merge

jamesjwu commented:
@pytorchbot rebase

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu commented:
@pytorchbot merge

jamesjwu committed:
Rebase

[ghstack-poisoned]

jamesjwu committed:
Rebase one more time

[ghstack-poisoned]

jamesjwu committed:
Fix merge conflict

[ghstack-poisoned]

jamesjwu committed:
Fix async compile

[ghstack-poisoned]

jamesjwu commented:
@pytorchbot merge

jamesjwu commented:
@pytorchbot  merge
